R/paste_linter.R

```diff
@@ -111,6 +123,76 @@ paste_linter <- function(allow_empty_sep = FALSE, allow_to_string = FALSE) {
     /parent::expr
   "
 
+  slash_str <- sprintf("STR_CONST[%s]", xp_text_in_table(c("'/'", '"/"')))
+  str_not_start_with_slash <-
+    "STR_CONST[not(substring(text(), 2, 1) = '/')]"
+  str_not_end_with_slash <-
+    "STR_CONST[not(substring(text(), string-length(text()) - 1, 1) = '/')]"
+  non_str <- "SYMBOL or expr"
+
+  # Type I: paste(..., sep = "/")
+  paste_file_path_xpath <- glue("
+  //SYMBOL_FUNCTION_CALL[text() = 'paste']
+    /parent::expr
+    /parent::expr[
+      SYMBOL_SUB[text() = 'sep']/following-sibling::expr[1][{slash_str}]
+      and not(SYMBOL_SUB[text() = 'collapse'])
+    ]
+  ")
+
+  # Type II: paste0(x, "/", y, "/", z)
```

Comment:
Due to the size of these comments, maybe you should interlace them with the XPath code.
NB `glue("a", "b") == "ab"` allowing for easy commenting - especially since the whitespace in the string isn't syntactic for XPath.

Reply:
I've been thinking about a wrapper that strips XPath "comments" for this, WDYT?

```r
xpath <- strip_comments("
  //path
    # we need this because...
    /expr
")
```

Reply:
in fact I would recommend using XPath2.0-style comments, in case we eventually have a back-end that supports XPath2.0:

https://stackoverflow.com/a/22648576/3576984

Reply:
IDE syntax highlighting and performance is likely better with R comments.
I'lm have to try and see if language injections for XPath highlight correctly in my IDE.
If we do put them in the string then definitely yes - use XPath 2.0 comment syntax.

Reply:
performance-wise, I'd expect a trivial difference (`strip_comments()` should just be a `gsub()` run in the linter factory).

will the difference for syntax highlighting be a big gain? as long as the comments are always on their own line, at least.

in favor of doing it within one big string literal b/c I find all the `"` and `,` for the approach of building the XPath line-by-line a bit distracting.

Reply:
PTAL

Reply:
Looks good.

Reply:
:sweat_smile:aesthetically it makes the code look a lot happier :upside_down_face: 🙂